### PR TITLE
[RFR] [BC Break] Upgrade react-router to v4

### PR DESCRIPTION
**Pull request by fzaninotto**
- [x] Make routing work
- [x] Make dashboard work
- [x] Make authentication work
- [x] Introduce `<Restricted>` component
- [x] Make custom routes work 
- [x] Update Custom App docs (BC Break: Custom app now radically different)
- [x] Update Custom Routes docs (BC Break: now taking an array of react-router v4 routes)
- [x] Update Custom Layout docs (BC Break)

Note: A v3 to v4 redux-form migration is NOT a piece of cake. A lot has changed, and the documentation is lacking, and the community resources are scarce.

### CODE REVIEW

**Comment by fzaninotto** [comment]
Switching to RFR

File: package.json [hunk 1]
```diff
@@ -62,12 +62,14 @@
     "lodash.set": "~4.3.2",
     "material-ui": "~0.16.5",
     "node-polyglot": "2.2.2",
+    "query-string": "^4.3.2",
```
Comment (djhi) [inline package.json]: multiple `^` spotted

**Comment by fzaninotto** [comment]
Review applied

File: docs/CustomApp.md [hunk 2]
```diff
@@ -47,39 +56,42 @@ const reducer = combineReducers({
...
-                    <CrudRoute path="comments" list={CommentList} create={CommentCreate} edit={CommentEdit} remove={Delete} />
-                    <CrudRoute path="users" list={UserList} create={UserCreate} edit={UserEdit} remove={Delete} />
-                </Route>
-            </Router>
+            <ConnectedRouter history={history}>
+                <MuiThemeProvider>
+                    <AppBar title="My Admin" />
+                    <Switch>
+                        <Route exact path="/" component={Dashboard} />
+                        <Route exact path="/posts" hasCreate render={(routeProps) => <PostList resource="posts" {...routeProps} />} />
```
Comment (JulienDemangeon) [inline docs/CustomApp.md]: We can maybe create a function to do routeProps injection?

Comment (fzaninotto) [inline docs/CustomApp.md]: No, this Custom App documentation is already a corner case that we don't want to emphasize.

**Pull request merged**